### PR TITLE
enhance: Add OGP data for notes with multiple images/videos

### PR DESCRIPTION
**Pull request by alexwh**
## What

Return all og:image or og:video meta tags based on the content of the note, rather than just the first match (`Array.find` vs `Array.filter`).

## Why

When misskey note links are embedded in external applications like Discord, only the first image was displayed.

## Additional info (optional)

[OGP supports multiple elements with arrays](https://ogp.me/#array).

I'm new to Pug and Misskey, so please suggest or make changes as needed (allow edits by maintainers is on).

## Checklist
- [x] Read the [contribution guide](https://github.com/misskey-dev/misskey/blob/develop/CONTRIBUTING.md)
- [ ] Test working in a local environment
- [ ] (If needed) Add story of storybook
- [x] (If needed) Update CHANGELOG.md
- [ ] (If possible) Add tests

### CODE REVIEW

File: packages/backend/src/server/web/views/note.pug [hunk 1]
```diff
@@ -19,15 +19,17 @@ block og
...
 	meta(property='og:description' content= summary)
 	meta(property='og:url'         content= url)
-	if video
-		meta(property='og:video:url'        content= video.url)
-		meta(property='og:video:secure_url' content= video.url)
-		meta(property='og:video:type'       content= video.type)
-		// FIXME: add width and height
-		// FIXME: add embed player for Twitter
-	if image
+	if videos
```
Comment (acid-chicken) [inline packages/backend/src/server/web/views/note.pug]:
```suggestion
	if videos.length
```

File: packages/backend/src/server/web/views/note.pug [hunk 2]
```diff
@@ -19,15 +19,17 @@ block og
...
+		each video in videos
+			meta(property='og:video:url'        content= video.url)
+			meta(property='og:video:secure_url' content= video.url)
+			meta(property='og:video:type'       content= video.type)
+			// FIXME: add width and height
+			// FIXME: add embed player for Twitter
+	if images
```
Comment (acid-chicken) [inline packages/backend/src/server/web/views/note.pug]:
```suggestion
	if images.length
```

**Review by acid-chicken — COMMENTED** [review]
maybe

**Comment by alexwh** [comment]
Good point, I didn't know that empty arrays were truthy until now (thanks javascript).

**Comment by tirr-c** [comment]
I've applied this patch to my server, but Discord still shows only the first one. Rendered pages were ok; there were multiple `og:image`'s when I checked it with curl.

**Comment by alexwh** [comment]
> I've applied this patch to my server, but Discord still shows only the first one. Rendered pages were ok; there were multiple `og:image`'s when I checked it with curl.

It seems Discord's implementation of Open Graph doesn't respect image arrays, and the only way to attach multiple images to an embed is [with a bot message](https://github.com/discord/discord-api-docs/issues/1643). I can't find any documentation on Discord's side for what properties are supported or how they're implemented, other than ["it's OGP"](https://twitter.com/discord/status/952834253842153472).

I think the only way to do this specifically for Discord Open Graph embeds would be to dynamically combine together the valid images into a single image with an image manipulation library and instead use that as the `og:image` tag. [BetterTwitFix](https://github.com/dylanpdx/BetterTwitFix) does this. Since the complexity of doing this is much higher compared to this original PR, I'll just leave it as an idea for now.

I'm not sure what other effects this PR has on other sites, but the maintainers can decide if this is still worth merging or not.

**Pull request merged**